README
Powermanga is an arcade 2D shoot-em-up game with 41 levels and more 
than 200 sprites. It runs in 320x200 or 640x400 pixels, 
with Window mode or full screen 
and support for 8, 15, 16, 24, and 32 bpp. 

PREREQUISITES

  You need:
  - A IA-32, x86-64, PowerPC, ARM or MIPS platform
  - 32 MB RAM and 80 MB swap
  - Linux 2.x 
  - SDL 1.2
  - SDL_mixer
  - the GCC C++ compiler
  - GNU make
  - (X.Org Server) 
  - (X11 with the MIT-SHM extension headers)
  - (LinuxThreads)

COMPILING AND INSTALLING


  To install in /usr/local :

       ./bootstrap
       ./configure
        make
        su -c 'make install'

LEGAL

  Powermanga is copyright TLK Games 1998-2012 and available under the
  terms of the GNU General Public License. See COPYING for details.


CHANGES
2015-01-10
Powermanga 0.93.1
- Fixes the speed of the stars after a new game that follows congratulation.
- Clears all gems after the final boss. Thanks to Josh Triplett for the report
  (Debian #764009)
- Disable insecure temporary file "/tmp/powermanga-log.txt". Thanks to 
  Josh Triplett for the report (Debian #764144)
- Translation of the man page in French.
- Fix file permissions in HTML directory. Thanks to  nemysisbsd for
  the report.
- Allows the joystick to be configured either from the command line or
  from the config file. Thanks to Maurerp for the patch.

2014-09-20
Powermanga 0.93
- Fix man page: The option to play the game in fullscreen mode
   must be --fullscreen not --full. (Patch Debian)
- Added keywords and a comment in German and French to the desktop file
  (Patch Debian)
- Fix configure.ac: Add custom CFLAGS (Patch Debian)
- Fix segmentation fault in the About menu (English version) 
- Fixes wrong joystick behaviour in display_sdl.c and allows the ship to move
  to the left side again. Thanks to Kalle Olavi Niemitalo for the report
  (Debian #561670)
- Allow compilation with Clang

2014-08-19
Powermanga 0.92
- Fix Debian bug #478213
- Fix configure.ac: replace AM_CONFIG_HEADER to AC_CONFIG_HEADERS
- Add  explicitly link with needed libm, fix Debian bug #632945
- Fix PNG_iTXt_SUPPORTED support
- Add italian language
- Fix negative coordinates explosions that caused a segmentation fault.

2012-08-26
Powermanga 0.91
- Update "configure.ac" file. 
- Fix compilation warnings (GCC 4.6.3 and Visual Studio)
- Fix alignment constraint (ARM and MIPS processors)
- Adds the ability to export all the game's graphics in PNG files
  using the command line.
- Build test (without sound support) with Visual Studio and run
  successfully on Windows Mobile (HTC Touch P3450) 
- Recognizes all the joysticks connected at startup
- Rewrites the file "music_game.zik" with Milkytracker.
  Now the library "SDL_mixer" can read and play this module.
- The application switches to pause when it loses focus.
- Minor bugfixes and improvements

2007-09-02
Powermanga 0.90
- the sounds were updated to replace non-free sounds
- joystick support was added
- scale2x, scale3x, and scale4x effects were added
- the order text was changed to support both plain text and keystrokes file
- the source code files were renamed
- the code was considerably cleaned up and rewritten
- comments and variables were translated to the English language
- all C++ code was removed and the remaining code was made to conform with C99
- the license was changed to the GPLv3
- resources files were renamed
- volume control was added
- GP2X and PSP patches were merged
- the high-score file format was changed

2006-03-27
Powermanga 0.80
- compile with GCC 4.0.3
- fix palette conversion in 16 bits RGB color mode (non-IA32 architecture)
- add config file
- add french language support
- various bugs fixes
- compile, run, and test the project on AMD64 (Ubuntu Dapper Drake) 
- compile, run, and test the project on Win32 (Patrice Duhamel)
- upload the source code on CVS (http://sourceforge.net/projects/powermanga/) 

2004-08-01:
Powermanga 0.79
- fix some minor bugs and memory leaks (Patrice Duhamel and Jozef Behran)
- duplicate set of cursor movement keys on the numeric keypad (modus)
- disable F key (full screen) if input player name

2003-07-14: 
Powermanga 0.78
- use automake/autoconf (by Samuel Hocevar)
- remplacing XFree routines by SDL routines
- game run in fullscreen

2003-07-01: (Major release)
Powermanga 0.77
- rewrite of the major part of the assembler code in langage C (Samuel Hocevar)
- port to Linux PowerPC (a fix for big endian problems)
- switching of SEAL routines to the SDL_mixer (musics and sounds)
- fix for bug in loading PCX files  (Samuel Hocevar)
- fix for bug of closing the game's window (with the close widget)

2003-04-07:
Powermanga 0.76 
 - bug of the 'uninstall' rules : 
   'rm -f -r $(EXEDIR)' instead 'rm -f -r $(EXEDIR)/$(PROG)'
   Thank to Torbjörn Andersson torbjorn.e.andersson@tietoenator.com

2003-03-29:
Powermanga 0.75 (Minor release)
- fix memory leak (courbe.cpp, gardiens.cpp and grille.cpp) and fix bug
  in linuxroutines.cpp by Samuel Hocevar <sam@zoy.org>
- fix bug in "playanim.cpp", write the AUTHORS file, and force by default
  in mode 640x480 pixels by Bruno Ethvignot <bruno@tlk.fr>

2002-08-20:
Powermanga 0.74 (Minor release)
- Update the man page, by Jeronimo Pellegrini <pellegrini@mpcnet.com.br>
- Update the scroll-text of the menu (given by Samuel Hocevar <sam@zoy.org>)
- Update Makefile (pacth given by Samuel Hocevar <sam@zoy.org>)

2002-08-17:
Powermanga 0.73 (Minor release)
- Fix a file loading bug. Don't find files in 
  "/usr/share/games/powermanga" (patch by Guillaume Cottenceau)
- Update Makefile

2002-08-11:
Powermanga 0.72 (Minor release)
- Ident the code with indent program and cleaning of the code 
- Fix thread music bug. Sometimes the sound did not work. 
  Because the thread would run in loop. This error prevented the program
  from leaving.
- Fix bug at the end of game. The game was blocked after the animation
  of the end. The player should press a key to continue.
- update Makefile (set optimizations options)
- Disable inline option in Deplacement_Tir_X() and Deplacement_Tir_Y()
  functions (patch by Guillaume Cottenceau, for gcc 3.1)
- Add --easy / --hard options (patch by Laurent Duperval)

2000-12-02:
Powermanga 0.71c (Minor release)
- Update Makefile
- Save score in /var/game/powermanga.hi
- Fixed files installing bugs 

2000-11-27:
Powermanga 0.71 (Initial release)

1999-08-17:
* start the Linux port


src/movie.c
/**
 * @file movie.c
 * @brief play start and congratulations animations files
 *        ("movie_congratulation.gca" and "movie_introduction.gca")
 * @date 2012-08-26 
 * @author Etienne Sobole
 */
/*
 * copyright (c) 1998-2015 TLK Games all rights reserved
 * $Id: movie.c,v 1.17 2012/08/26 17:09:14 gurumeditation Exp $
 *
 * Powermanga is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 3 of the License, or
 * (at your option) any later version.
 *
 * Powermanga is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston,
 * MA  02110-1301, USA.
 */
#include "config.h"
#include "powermanga.h"
#include "tools.h"
#include "assembler.h"
#include "tools.h"
#include "config_file.h"
#include "display.h"
#include "log_recorder.h"
#include "movie.h"

/** Wich movie si played: 1=intro or 2=congratulation */
Uint32 movie_playing_switch = MOVIE_INTRODUCTION;
/** Pointer to the buffer for the current animation movie */
unsigned char *movie_buffer = NULL;
static unsigned char *image2 = NULL;
static unsigned char *icmpr = NULL;
static unsigned char *smage1 = NULL;
static unsigned char *smage2 = NULL;
static unsigned char *scmpr = NULL;
/** Pointer to the filedata of the movie in memory */
static unsigned char *movie_filedata = NULL;
/* pointer to the colormap in the file */
static Sint32 movie_counter = 0;
static Sint32 images = 0;

static bool movie_initialize (const char *filename);
static bool movie_load (const char *filename);
static bool movie_play (void);
static unsigned char *decompress (unsigned char *, unsigned char *,
                                  unsigned char *);

/**
 * Play movie animation compressed
 * @return TRUE if it completed successfully or 0 otherwise
 */
bool
movie_player (void)
{
  const char *filename;
  switch (movie_playing_switch)
    {
      /* introduction movie */
    case MOVIE_INTRODUCTION:
      filename = "graphics/movie_introduction.gca";
      if (movie_initialize (filename))
        {
          movie_playing_switch = MOVIE_PLAYED_CURRENTLY;
        }
      else
        {
          movie_playing_switch = MOVIE_NOT_PLAYED;
          LOG_ERR ("movie_initialize(\"%s\") failed", filename);
          return FALSE;
        }
      break;
      /* congratulation movie */
    case MOVIE_CONGRATULATIONS:
      filename = "graphics/movie_congratulation.gca";
      if (movie_initialize (filename))
        {
          movie_playing_switch = MOVIE_PLAYED_CURRENTLY;
        }
      else
        {
          movie_playing_switch = MOVIE_NOT_PLAYED;
          LOG_ERR ("movie_initialize(\"%s\") failed", filename);
          return FALSE;
        }
      break;
      /* play movie animation compressed */
    case MOVIE_PLAYED_CURRENTLY:
      {
        if (!movie_play () || key_code_down > 0 || fire_button_down
            || mouse_b > 1)
          {
            movie_free ();
            movie_playing_switch = MOVIE_NOT_PLAYED;
          }
      }
      break;
    }
  return TRUE;
}

/**
 * Initialize an animation file, allocate buffers and create offscreen
 * @param filename: the file which should be loaded.
 * @return boolean value on success or failure
 */
static bool
movie_initialize (const char *filename)
{
  Sint32 i = 0;
  smage1 = smage2 = scmpr = NULL;
  smage1 = ((unsigned char *) memory_allocation (64000));
  if (smage1 == NULL)
    {
      LOG_ERR ("not enough memory to allocate 'smage1'");
      return FALSE;
    }
  smage2 = ((unsigned char *) memory_allocation (64000));
  if (smage1 == NULL)
    {
      LOG_ERR ("not enough memory to allocate 'smage2'");
      return FALSE;
    }
  movie_filedata = ((unsigned char *) memory_allocation ((1024 * 1024 * 4)));
  if (movie_filedata == NULL)
    {
      LOG_ERR ("not enough memory to allocate 'movie_filedata'");
      return FALSE;
    }
  for (i = 0; i < 64000; i++)
    {
      smage1[i] = 0;
      smage2[i] = 0;
    }
  if (!movie_load (filename))
    {
      LOG_ERR ("movie_load(%s) failed!", filename);
      return FALSE;
    }
  movie_buffer = smage1;
  image2 = smage2;
  icmpr = scmpr;
  movie_counter = 0;
  if (!create_movie_offscreen ())
    {
      LOG_ERR ("create_movie_buffer() failed!");
      return FALSE;
    }
  return TRUE;
}

/**
 * Deallocates the memory used by the movie player
 */
void
movie_free (void)
{
  if (smage1 != NULL)
    {
      free_memory ((char *) smage1);
      smage1 = NULL;
    }
  if (smage2 != NULL)
    {
      free_memory ((char *) smage2);
      smage2 = NULL;
    }
  if (movie_filedata != NULL)
    {
      free_memory ((char *) movie_filedata);
      movie_filedata = NULL;
    }
  if (pal16PlayAnim != NULL)
    {
      free_memory ((char *) pal16PlayAnim);
      pal16PlayAnim = NULL;
    }
  if (pal32PlayAnim != NULL)
    {
      free_memory ((char *) pal32PlayAnim);
      pal32PlayAnim = NULL;
    }
  destroy_movie_offscreen ();
}

/**
 * Load an animation file and convert colors palette
 * @param filename: the file which should be loaded.
 * @return boolean value on success or failure
 */
static bool
movie_load (const char *filename)
{
  Sint32 i, *ptr32;
  unsigned char *_p8, *_p, *_pPal, *pcxpal;
  if (!loadfile_into_buffer (filename, (char *) movie_filedata))
    {
      return FALSE;
    }
  ptr32 = (Sint32 *) movie_filedata;
  images = little_endian_to_int (ptr32++);
  _p8 = (unsigned char *) ptr32;
  pcxpal = _p8;
  scmpr = _p8 + 768;
  if (bytes_per_pixel == 2)
    {
      if (pal16PlayAnim == NULL)
        {
          pal16PlayAnim = (Uint16 *) memory_allocation (256 * 2);
          if (pal16PlayAnim == NULL)
            {
              LOG_ERR ("not enough memory to allocate 'pal16PlayAnim'");
              return FALSE;
            }
        }
      convert_palette_24_to_16 (pcxpal, pal16PlayAnim);
    }
  if (bytes_per_pixel > 2)
    {
      if (pal32PlayAnim == NULL)
        {
          pal32PlayAnim = (Uint32 *) memory_allocation (256 * 4);
          if (pal32PlayAnim == NULL)
            {
              LOG_ERR ("not enough memory to allocate 'pal32PlayAnim'");
              return FALSE;
            }
        }
      _p = (unsigned char *) pal32PlayAnim;
      _pPal = pcxpal;
      for (i = 0; i < 256; i++)
        {
#if SDL_BYTEORDER == SDL_BIG_ENDIAN
          _p[3] = _pPal[2];
          _p[2] = _pPal[1];
          _p[1] = _pPal[0];
          _p[0] = 0;
#else
          if (power_conf->scale_x >= 2)
            {
              _p[0] = _pPal[2];
              _p[1] = _pPal[1];
              _p[2] = _pPal[0];
              _p[3] = 0;
            }
          else
            {
              _p[2] = _pPal[2];
              _p[1] = _pPal[1];
              _p[0] = _pPal[0];
              _p[3] = 0;
            }
#endif
          _p += 4;
          _pPal += 3;
        }
    }
  return TRUE;
}

/*
 * Play a movie
 * @return TRUE if the movie is finished
 */
static bool
movie_play (void)
{
  /* images = 64 + 16 + 2 + 1 + 1 */
  unsigned char *tmp;
  movie_counter++;
  if (movie_counter == images - 1)
    {
      return FALSE;
    }
  icmpr = decompress (icmpr, movie_buffer, image2);
  tmp = movie_buffer;
  movie_buffer = image2;
  image2 = tmp;
  return TRUE;
}

/*
 *
 */
static unsigned char *
decompress (unsigned char *wsc, unsigned char *im1, unsigned char *im2)
{
  Sint32 i, j, mode;
  Uint32 wr;
  unsigned char *idec;
  unsigned char *wdec;
  unsigned char c;
  unsigned char color;
  Sint32 length = 0;
  Sint32 retour = 0;
  Sint32 position = 0;
  i = 0;
  idec = im2;
  while (i < 64000)
    {
      mode = 0;
      /* read a byte */
      c = *wsc++;
      if (c == 255)
        {
          color = *wsc++;
          *idec++ = color;
          i++;
          mode = 0;
        }
      else if ((c & 0xc0) == 0)
        {
          wr = 0;
          wr = c << 16;
          wr += (*wsc++) << 8;
          wr += *wsc++;
          length = wr & 63;
          position = (wr >> 6) & 65535;
          mode = 1;
        }
      else if ((c & 0xc0) == 0x40)
        {
          wr = 0;
          wr = c << 24;
          wr += (*wsc++) << 16;
          wr += (*wsc++) << 8;
          wr += *wsc++;
          length = wr & 16383;
          position = (wr >> 14) & 65535;
          mode = 1;
        }
      else if ((c & 0xe0) == 0x80)
        {
          wr = 0;
          wr = c << 8;
          wr += *wsc++;
          length = wr & 63;
          retour = (wr >> 6) & 255;
          mode = 2;
        }
      else if ((c & 0xe0) == 0xa0)
        {
          wr = 0;
          wr = c << 16;
          wr += (*wsc++) << 8;
          wr += *wsc++;
          length = wr & 255;
          retour = (wr >> 8) & 8191;
          mode = 2;
        }
      else if ((c & 0xe0) == 0xc0)
        {
          wr = 0;
          wr = c << 24;
          wr += (*wsc++) << 16;
          wr += (*wsc++) << 8;
          wr += *wsc++;
          length = wr & 8191;
          position = (wr >> 13) & 65535;
          mode = 3;
        }
      if (i + length > 64000)
        {
          length = 64000 - i;
        }
      if (mode == 1)
        {
          for (j = 0; j < length; j++)
            {
              *idec++ = im1[position + j];
            }
          i += length;
        }
      else if (mode == 2)
        {
          wdec = idec - retour;
          for (j = 0; j < length; j++)
            {
              *idec++ = *wdec++;
            }
          i += length;
        }
      else if (mode == 3)
        {
          for (j = 0; j < length; j++)
            {
              *idec++ = im2[position + j];
            }
          i += length;
        }
    }
  return wsc;
}
